ReplayMessenger.base.lib.nut

```diff
@@ -0,0 +1,481 @@
+                // Send acknowledgment automatically
+                error = _partner.send(RM_MESSAGE_TYPE_ACK, {
+                    "id" : id
+                });
+            }
+        } else {
+            // No valid handler
+            _log("No valid handler. Id: " + id);
+        }
+
+        if (error) {
+            _log("Ack failed. Id: " + id);
+        }
+    }
+
+    // Calls the error handler if set for the message with the specified error
+    function _onSendFail(msg, error) {
+        local id = msg.payload.id;
+        _log("Failed to send or haven't received Ack. Id: " + id + " Reason: " + error);
+
+        if (id in _sentQueue) {
+            delete _sentQueue[id];
+        }
+
+        if (_isFunction(_onFail)) {
+            _onFail(msg, error);
+        }
+    }
+
+    // Processes _sentQueue
+    function _processQueues() {
+        // Clean up the timer
+        _queueTimer = null;
+
+        local now = time();
+
+        // Call onFail for timed out messages
+        foreach (id, msg in _sentQueue) {
+            local ackTimeout = msg._ackTimeout ? msg._ackTimeout : _ackTimeout;
+            if (now - msg._sentTime >= ackTimeout) {
+                _onSendFail(msg, RM_ERR_ACK_TIMEOUT);
+            }
+        }
+
+        // Restart the timer if there is something pending
+        if (!_isAllProcessed()) {
+            _setTimer();
+        }
+    }
+
+    // Incremental message id generator
+    function _getNextId() {
+        _nextId %= RAND_MAX;
+        return _nextId++;
+    }
+
+    // Returns true if there are no messages to process (_sentQueue is empty)
+    function _isAllProcessed() {
+        return _sentQueue.len() == 0;
+    }
+
+    // Returns true if we are on the agent
+    function _isAgent() {
+        return imp.environment() == ENVIRONMENT_AGENT;
+    }
+
+    // Sets a timer for processing queues
+    function _setTimer() {
+        if (_queueTimer) {
+            // The timer is already running
+            return;
+        }
+        _queueTimer = imp.wakeup(RM_QUEUE_CHECK_INTERVAL_SEC,
+                                _processQueues.bindenv(this));
+    }
+
+    // Returns true if the argument is a function and false otherwise
+    function _isFunction(f) {
+        return f && typeof f == "function";
+    }
+
+    // Implements debug logging. Sends the log message to the console output if "debug" configuration flag is set
+    function _log(message) {
+        if (_debug) {
+            server.log("[RM] " + message);
+        }
+    }
+
+    // Returns the current value of the "monotonic" millisecond timer
+    //
+    // NOTE: the timer can overflow!!!
+    // The primary purpose is rate measurements, where the above
+    // limitation is not critical.
+    function _monotonicMillis() {
+        return _isAgent() ? time() * 1000 + date().usec / 1000 : hardware.millis();
+    }
+
+    function _typeof() {
+        return "ReplayMessengerBase";
+    }
+}
```

Comment:
It's better to have empty lines at the end. I remember there were some edge cases triggered by this in Builder and on the impCentral sides )